performance-tracking-core/src/main/java/com/rakuten/tech/mobile/perf/core/Sender.java

```diff
@@ -27,65 +27,53 @@
    *
-   * @param startIndex send measurements starting from this index (of the buffer)
-   * @return next unsent index, i.e. {@code startIndex} for the next call to send
+   * @param startId send measurements starting from this Id (of the buffer)
+   * @return next unsent Id, i.e. {@code startId} for the next call to send
    */
-  int send(int startIndex) throws IOException {
-    int endIndex = buffer.nextTrackingId.get() % MeasurementBuffer.SIZE;
-    if (endIndex < 0) {
-      endIndex += MeasurementBuffer.SIZE;
-    }
-
-    int count;
-
-    if (startIndex == endIndex && buffer.next() == null) {
-      count = MeasurementBuffer.SIZE;
-    } else {
-      count = endIndex - startIndex;
-      if (count < 0) {
-        count += MeasurementBuffer.SIZE;
-      }
-    }
+  int send(int startId) throws IOException {
+    int count = buffer.count(startId);
 
     if (count >= MIN_COUNT) {
-      startIndex = send(startIndex, count);
+      startId = send(startId, count);
+    } else {
+      sendSingleMetric();
     }
 
-    return startIndex;
+    return startId;
   }
 
   /**
-   * Tries to send measurements from measurement buffer starting at {@code startIndex} until
-   * {@code endIndex}. Will stop sending if the metric/measurement is potentially still "alive",
+   * Tries to send measurements from measurement buffer starting at {@code startId} until
+   * {@code count} is reached. Will stop sending if the metric/measurement is potentially still "alive",
    * i.e. it started less than {@link Metric#MAX_TIME} or {@link Measurement#MAX_TIME} ago
-   * respectively. Returns the buffer index next unsent measurement/metric, which should be
-   * used for the next call to this method as new {@code startIndex}.
+   * respectively. Returns the Id for next unsent measurement/metric, which should be
+   * used for the next call to this method as new {@code startId}.
    *
-   * @param startIndex send measurements starting from this index (of the buffer)
+   * @param startId send measurements starting from this Id (of the buffer)
    * @param count      send `count` many measurements
-   * @return last sent index + 1, i.e. {@code startIndex} for the next call to send
+   * @return last sent Id + 1, i.e. {@code startId} for the next call to send
    */
-  private int send(int startIndex, int count) throws IOException {
+  private int send(int startId, int count) throws IOException {
     sent = 0;
     long now = System.currentTimeMillis();
     Metric savedMetric = metric == null ? null : metric.copy();
 
     try {
-      int endIndex = (startIndex + count) % MeasurementBuffer.SIZE;
-      for (int i = startIndex; count > 0; count--, i = (i + 1) % MeasurementBuffer.SIZE) {
-        Measurement m = buffer.at[i];
+      int endId = startId + count;
+      for (int i = startId; i < endId; i++) {
+        Measurement m = buffer.getByTrackingId(i);
 
-        if (m.type == Measurement.METRIC) {
+        if (m == null) {
+          continue;
```

Comment:
It shouldn't ever be able to happen, but technically `MeasurementBuffer#getByTrackingId` can return null. But actually, I've changed this back to using the index instead of ID, so this isn't relevant anymore.